src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.Fixer.cs

```diff
@@ -0,0 +1,173 @@
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic), Shared]
+    public sealed class StringBuilderAppendShouldNotTakeSubstringFixer : CodeFixProvider
+    {
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray
+            .Create(
+                StringBuilderAppendShouldNotTakeSubstring.RuleIdOneParameterId,
+                StringBuilderAppendShouldNotTakeSubstring.RuleIdTwoParameterId);
+
+        public sealed override FixAllProvider GetFixAllProvider()
+        {
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            SyntaxNode root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            SyntaxNode node = root.FindNode(context.Span, getInnermostNodeForTie: true);
+
+            foreach (var diagnostic in context.Diagnostics)
+            {
+                if (diagnostic.Id == StringBuilderAppendShouldNotTakeSubstring.RuleIdOneParameterId)
+                {
+                    context
+                        .RegisterCodeFix(
+                            new MyCodeAction(
+                                MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadWithOneParameterFix,
+                                ctx => FixCodeOneParameter(context.Document, root, node, ctx),
+                                equivalenceKey: MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadWithOneParameterFix),
+                            diagnostic);
+                }
+                else
+                {
+                    context
+                        .RegisterCodeFix(
+                            new MyCodeAction(
+                                MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadWithTwoParameterFix,
+                                ctx => FixCodeTwoParameters(context.Document, root, node, ctx),
+                                equivalenceKey: MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadWithTwoParameterFix),
+                            diagnostic);
+                }
+            }
+        }
+
+        private static async Task<Document> FixCodeOneParameter(
+            Document document,
+            SyntaxNode root,
+            SyntaxNode nodeToFix,
+            CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var typedNodeToFix =  semanticModel.GetOperation(nodeToFix) as IInvocationOperation;
+            var generator = SyntaxGenerator.GetGenerator(document);
```

Comment:
what happens if .GetOperation doesn't return an IInvocationOperation?

Reply:
that's a good question.
Implicitly that should never happen as long as the fixer is coupled to the Analyzer, as the corresponding check is part of the analyzer and necessary to report the diagnostics.
Of course that's not stable to rely on that.
What's the best way to deal with it then? Just do nothing if it's not an IInvocationOperation? throw an exception?
For now I'll just do nothing and return the document. Would like to know what's the best approach.

Reply:
best  thing is in the RegisterCodeFixes function to check that the info from the diagnostic is sufficient to get by.  

Note: i don't believe this is true:

> Implicitly that should never happen as long as the fixer is coupled to the Analyzer, as the corresponding check is part of the analyzer and necessary to report the diagnostics.

Just because you reported a diagnostic on a certain span doesn't mean that span will map back to that IOperation you started with.  A great example of that is how you can trivially make that happen if you dn't pass `getInnermostNodeForTie: true`.  But, in general, there is no guarantee about what IOp you get back for a particular syntax node.  So you have to assume you may get nothing back, or you may get something different.  So it's best to check first, verify you get the right thing back, then actually register the fixes.